### PR TITLE
PERF: Make __iadd__ actually inplace

### PR DESCRIPTION
**Pull request by lithomas1**
- [ ] closes #xxxx (Replace xxxx with the GitHub issue number)
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

Starting simple with just ``__iadd__`` for now, eventually want to expand to all arith/logical ops? when I'm feeling braver :).

Everything except:
- SparseArray
- MaskedArray(maybe just needs tests?)
- Some special case with DF+Series (I can't seem to hit this path normally, need to investigate more)

should be implemented. I'll try to get those in a follow up.

Hopefully, I'm not using abusing/breaking existing frame/manager plumbing. I've kinda been hackily inserting return self calls all over.

### CODE REVIEW

**Review by phofl — CHANGES_REQUESTED** [review]
Can you add tests for copy on write?

File: pandas/core/arrays/sparse/array.py [hunk 1]
```diff
@@ -1701,7 +1702,12 @@ def __array_ufunc__(self, ufunc: np.ufunc, method: str, *inputs, **kwargs):
     # Ops
     # ------------------------------------------------------------------------
 
-    def _arith_method(self, other, op):
+    def _arith_method(self, other, op, inplace: bool = False):
+        if op in _inplace_ops:
+            # TODO: Do ops actually inplace on SparseArray
+            # Replace the inplace op with the normal op
+            op = _inplace_ops[op]  #
```
Comment (jbrockmendel) [inline pandas/core/arrays/sparse/array.py:1709]: extra trailing `#`?

File: pandas/core/computation/expressions.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def set_numexpr_threads(n=None) -> None:
         ne.set_num_threads(n)
 
 
-def _evaluate_standard(op, op_str, a, b):
+def _evaluate_standard(op, op_str, a, b, inplace):
```
Comment (jbrockmendel) [inline pandas/core/computation/expressions.py:67]: can annotate inplace here?  (probably also op_str?)

File: pandas/core/computation/expressions.py [hunk 3]
```diff
@@ -128,13 +129,21 @@ def _evaluate_numexpr(op, op_str, a, b):
...
     if result is None:
-        result = _evaluate_standard(op, op_str, a, b)
+        result = _evaluate_standard(op, op_str, a, b, inplace)
 
     return result
 
 
+# Inplace ops and their fallbacks, in-case
+# the operation actually can't be done inplace
+_inplace_ops = {operator.iadd: operator.add}
```
Comment (jbrockmendel) [inline pandas/core/computation/expressions.py:139]: is the idea to eventually do other besides add?

File: pandas/core/computation/expressions.py [hunk 4]
```diff
@@ -234,13 +243,27 @@ def evaluate(op, a, b, use_numexpr: bool = True):
     b : right operand
     use_numexpr : bool, default True
         Whether to try to use numexpr.
+    inplace: bool, default False
+        Whether to do the op inplace.
+        If False, will replace
+        inplace op with the non-inplace version of it.
+        a must by an ndarray/EA in this case
```
Comment (jbrockmendel) [inline pandas/core/computation/expressions.py:250]: maybe ticks around "a"?

File: pandas/core/computation/expressions.py [hunk 5]
```diff
@@ -234,13 +243,27 @@ def evaluate(op, a, b, use_numexpr: bool = True):
...
     """
     op_str = _op_str_mapping[op]
+    inplace = inplace and op in _inplace_ops
+    if inplace:
+        inplace = np.result_type(a, b) == a.dtype
```
Comment (jbrockmendel) [inline pandas/core/computation/expressions.py:255]: docstring says `a` can be an EA, but id expect that to raise in np.result_type.  am i wrong there? 

come to think of it, is expect numexpr to choke on EAs in general

File: pandas/core/frame.py [hunk 6]
```diff
@@ -7420,23 +7421,35 @@ def _cmp_method(self, other, op):
...
         if self._should_reindex_frame_op(other, op, 1, None, None):
             return self._arith_method_with_reindex(other, op)
 
         axis: Literal[1] = 1  # only relevant for Series other case
         other = ops.maybe_prepare_scalar_for_op(other, (self.shape[axis],))
 
-        self, other = self._align_for_op(other, axis, flex=True, level=None)
+        # Disable copy on write, since align will make a shallow copy
+        # (preventing it from going inplace when it should be able to)
+        # TODO: better way of doing this?
```
Comment (jbrockmendel) [inline pandas/core/frame.py:7433]: cc @phofl?

File: pandas/core/internals/managers.py [hunk 7]
```diff
@@ -322,6 +323,10 @@ def apply(
         f : str or callable
             Name of the Block method to apply.
         align_keys: List[str] or None, default None
+        block_inplace:
+            Whether to do the operation inplace.
+            Note: This will always obey Copy on Write rules
+            (so a black sharing references will never be operated inplace on)
```
Comment (jbrockmendel) [inline pandas/core/internals/managers.py:329]: typo black->block

Comment (jbrockmendel) [reply pandas/core/internals/managers.py:329]: "operated inplace on" -> "operated upon inplace"?

File: pandas/core/internals/managers.py [hunk 8]
```diff
@@ -338,6 +343,10 @@ def apply(
         aligned_args = {k: kwargs[k] for k in align_keys}
 
         for b in self.blocks:
+            if block_inplace:
+                # Check if inplace allowed according to CoW
+                if using_copy_on_write() and b.refs.has_reference():
```
Comment (jbrockmendel) [inline pandas/core/internals/managers.py:348]: call using_copy_on_write once outside the loop?

File: pandas/core/frame.py (hunk 6 above)
Comment (phofl) [reply pandas/core/frame.py:7433]: Yeah we definitely shouldn't do this. This shouldn't matter anyway? You can't do it inlace if the index of left changed and in all other cases you can simply discard left, freeing the reference

File: pandas/core/internals/managers.py (hunk 8 above)
Comment (phofl) [reply pandas/core/internals/managers.py:348]: This shouldn't live here. These logic live on the block level or rather in ``operate_blockwise`` in this case.

**Comment by mroeschke** [comment]
Closing to clear the queue, but feel free to reopen when you have time to revisit

**Pull request closed without merging**